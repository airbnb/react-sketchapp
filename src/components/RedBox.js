/* @flow */
import React from 'react';
import PropTypes from 'prop-types';
import ErrorStackParser from 'error-stack-parser';
import Text from './Text';
import View from './View';

type StackFrame = {
  functionName: string,
  isConstrutor?: boolean,
  isEval?: boolean,
  isNative?: boolean,
  isTopLevel?: boolean,
  columnNumber?: number,
  lineNumber?: number,
  fileName?: string,
  functionName?: string,
  source?: string,
  args?: any[],
  evalOrigin?: StackFrame
};

const styles = {
  redbox: {
    padding: 10,
    width: 480,
    backgroundColor: 'rgb(204, 0, 0)',
  },
  frame: {},
  message: {
    fontWeight: 'bold',
    fontSize: 16,
    lineHeight: 16 * 1.2,
    color: 'white',
  },
  stack: {
    fontFamily: 'Monaco',
    marginTop: 20,
    color: 'white',
  },
};

const propTypes = {
  error: PropTypes.oneOfType([PropTypes.instanceOf(Error), PropTypes.string])
    .isRequired,
  // filename: PropTypes.string,
  // editorScheme: PropTypes.string,
  // useLines: PropTypes.bool,
  // useColumns: PropTypes.bool,
};

class RedBox extends React.Component {
  static defaultProps = {
    useLines: true,
    useColumns: true,
  };

  renderFrames(frames: Array<StackFrame>) {
    return frames.map((f, index) => (
      <Text key={index} style={styles.stack}>
        {f.functionName}
      </Text>
    ));
  }

  render() {
    const { error } = this.props;

    if (typeof error === 'string') {
      return (
        <View name="RedBox" style={styles.redbox}>
          <Text name="Message" style={styles.message}>{`Error: ${error}`}</Text>
        </View>
      );
    }

    let frames;
    let parseError;

    let frameChildren;

    try {
      frames = ErrorStackParser.parse(error);
    } catch (e) {
      parseError = new Error(
        'Failed to parse stack trace. Stack trace information unavailable.'
      );
    }

    if (parseError) {
      frameChildren = (
        <View style={styles.frame} key={0}>
          <View>{parseError.message}</View>
        </View>
      );
    }

    if (frames) {
      frameChildren = this.renderFrames(frames);
    }

    return (
      <View name="RedBox" style={styles.redbox}>
        <Text name="Message" style={styles.message}>{`${error.name}: ${
          error.message
        }`}</Text>
        <View name="Frames" style={styles.stack}>
          {frameChildren}
        </View>
      </View>
    );
  }
}

RedBox.propTypes = propTypes;

export default RedBox;
