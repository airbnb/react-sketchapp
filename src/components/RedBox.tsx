import * as React from 'react';
import * as PropTypes from 'prop-types';
import ErrorStackParser from 'error-stack-parser';
import Text from './Text';
import View from './View';

type StackFrame = {
  isConstrutor?: boolean;
  isEval?: boolean;
  isNative?: boolean;
  isTopLevel?: boolean;
  columnNumber?: number;
  lineNumber?: number;
  fileName?: string;
  functionName?: string;
  source?: string;
  args?: any[];
  evalOrigin?: StackFrame;
};

const styles = {
  redbox: {
    padding: 10,
    width: 480,
    backgroundColor: 'rgb(204, 0, 0)',
  },
  frame: {},
  message: {
    fontWeight: 'bold',
    fontSize: 16,
    lineHeight: 16 * 1.2,
    color: 'white',
  },
  stack: {
    fontFamily: 'Monaco',
    marginTop: 20,
    color: 'white',
  },
};

export const ErrorBoxPropTypes = {
  error: PropTypes.oneOfType([PropTypes.instanceOf(Error), PropTypes.string]).isRequired,
  // filename: PropTypes.string,
  // editorScheme: PropTypes.string,
  // useLines: PropTypes.bool,
  // useColumns: PropTypes.bool,
};

type Props = PropTypes.InferProps<typeof ErrorBoxPropTypes>;

export default class RedBox extends React.Component<Props> {
  static propTypes = ErrorBoxPropTypes;

  static defaultProps = {
    useLines: true,
    useColumns: true,
  };

  renderFrames(frames: Array<StackFrame>) {
    return frames.map((f, index) => (
      <Text key={index} style={styles.stack}>
        {f.functionName}
      </Text>
    ));
  }

  render() {
    const { error } = this.props;

    if (typeof error === 'string') {
      return (
        <View name="RedBox" style={styles.redbox}>
          <Text name="Message" style={styles.message}>{`Error: ${error}`}</Text>
        </View>
      );
    }

    let frames: ErrorStackParser.StackFrame[] | undefined;
    let parseError: Error | undefined;

    let frameChildren: JSX.Element[] | JSX.Element | undefined;

    try {
      frames = ErrorStackParser.parse(error);
    } catch (e) {
      parseError = new Error('Failed to parse stack trace. Stack trace information unavailable.');
    }

    if (parseError) {
      frameChildren = (
        <View style={styles.frame} key={0}>
          <View>{parseError.message}</View>
        </View>
      );
    }

    if (frames) {
      frameChildren = this.renderFrames(frames);
    }

    return (
      <View name="RedBox" style={styles.redbox}>
        <Text name="Message" style={styles.message}>
          {`${error.name}: ${error.message}`}
        </Text>
        <View name="Frames" style={styles.stack}>
          {frameChildren}
        </View>
      </View>
    );
  }
}
