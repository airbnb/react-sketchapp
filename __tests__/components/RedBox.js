import React from 'react';
import renderer from 'react-test-renderer';
import RedBox from '../../src/components/RedBox';

describe('<RedBox />', () => {
  it.skip('renders simple errors', () => {
    const tree = renderer.create(<RedBox error={new Error('THIS IS AN ERROR')} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders string errors', () => {
    const tree = renderer.create(<RedBox error="String only error" />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
