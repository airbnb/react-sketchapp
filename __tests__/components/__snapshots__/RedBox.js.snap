exports[`<RedBox /> renders simple errors 1`] = `
<view
  name="RedBox"
  resizingConstraint={undefined}
  style={
    Object {
      "backgroundColor": "rgb(204, 0, 0)",
      "paddingBottom": 10,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 10,
      "width": 480,
    }
  }>
  <text
    name="Message"
    resizingConstraint={undefined}
    style={
      Object {
        "color": "white",
        "fontSize": 16,
        "fontWeight": "bold",
        "lineHeight": 19.2,
      }
    }>
    Invariant Violation: THIS IS AN ERROR
  </text>
  <view
    name="Frames"
    resizingConstraint={undefined}
    style={
      Object {
        "color": "white",
        "fontFamily": "Monaco",
        "marginTop": 20,
      }
    }>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      invariant
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      Object.<anonymous>
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      Object.<anonymous>
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      attemptAsync
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      QueueRunner.run
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      } />
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      } />
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      Object.fn
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      attemptAsync
    </text>
    <text
      name="Text"
      resizingConstraint={undefined}
      style={
        Object {
          "color": "white",
          "fontFamily": "Monaco",
          "marginTop": 20,
        }
      }>
      QueueRunner.run
    </text>
  </view>
</view>
`;

exports[`<RedBox /> renders string errors 1`] = `
<view
  name="RedBox"
  resizingConstraint={undefined}
  style={
    Object {
      "backgroundColor": "rgb(204, 0, 0)",
      "paddingBottom": 10,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 10,
      "width": 480,
    }
  }>
  <text
    name="Message"
    resizingConstraint={undefined}
    style={
      Object {
        "color": "white",
        "fontSize": 16,
        "fontWeight": "bold",
        "lineHeight": 19.2,
      }
    }>
    Error: String only error
  </text>
</view>
`;
